File: sales_performance/sales_performance/report/sales_target_achievement/sales_target_achievement.js
frappe.query_reports["Sales Target Achievement"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "fiscal_year",
			label: __("Fiscal Year"),
			fieldtype: "Link",
			options: "Fiscal Year",
			reqd: 1,
			default: frappe.defaults.get_user_default("fiscal_year")
				|| (frappe.boot.current_fiscal_year && frappe.boot.current_fiscal_year[0]),
		},
		{
			fieldname: "pay_on",
			label: __("Pay Incentive On"),
			fieldtype: "Select",
			options: "Amount\nQty",
			default: "Amount",
			reqd: 1,
			description: __("The earned salesperson payout is shared across items in proportion to their positive Qty or Amount surplus."),
		},
		{
			fieldname: "period",
			label: __("Period"),
			fieldtype: "Select",
			options: ["Annual", "Monthly", "Quarterly"],
			default: "Monthly",
		},
		{ fieldname: "month", label: __("Month"), fieldtype: "Int", depends_on: "eval:doc.period === 'Monthly'" },
		{ fieldname: "quarter", label: __("Quarter"), fieldtype: "Int", depends_on: "eval:doc.period === 'Quarterly'" },
		{
			fieldname: "incentive_band",
			label: __("Incentive Achievement"),
			fieldtype: "Select",
			options: "All\nMin\nMax\nNot Achieve",
			default: "All",
			description: __("Min/Max show the earned incentive band. Not Achieve shows rows with no allocated incentive."),
		},
		{ fieldname: "sales_person", label: __("Sales Person"), fieldtype: "Link", options: "Sales Person" },
		{ fieldname: "territory", label: __("Territory"), fieldtype: "Link", options: "Territory" },
		{ fieldname: "item_group", label: __("Item Group"), fieldtype: "Link", options: "Item Group",
			get_query() { return { query: "sales_performance.api.planning.item_group_query" }; } },
		{ fieldname: "customer_group", label: __("Customer Group"), fieldtype: "Link", options: "Customer Group" },
		{ fieldname: "item", label: __("Item"), fieldtype: "Link", options: "Item" },
		{ fieldname: "from_date", label: __("From Date"), fieldtype: "Date" },
		{ fieldname: "to_date", label: __("To Date"), fieldtype: "Date" },
	],
	formatter(value, row, column, data, default_formatter) {
		if (data && data.is_total_row) {
			return `<strong>${default_formatter(value, row, column, data)}</strong>`;
		}
		value = (window.sales_performance && sales_performance.report_formatter)
			? sales_performance.report_formatter(value, row, column, data, default_formatter)
			: default_formatter(value, row, column, data);
		if (!data) {
			return value;
		}
		if (column.fieldname === "incentive_on_amount" && flt(data.incentive_on_amount) > 0) {
			value = `<span style="color:#1a5276">${value}</span>`;
		}
		if (column.fieldname === "incentive_on_qty" && flt(data.incentive_on_qty) > 0) {
			value = `<span style="color:#6c3483">${value}</span>`;
		}
		if (column.fieldname === "incentive_amount" && flt(data.incentive_amount) > 0) {
			value = `<span style="font-weight:600;color:#16833b">${value}</span>`;
		}
		return value;
	},
};

File: sales_performance/sales_performance/report/sales_target_achievement/sales_target_achievement.py
import frappe
from frappe import _

from sales_performance.services.achievement_engine import compute_metrics
from sales_performance.services.numbers import nflt
from sales_performance.services.incentive_engine import (
	allocate_item_incentives,
	collect_achievement_rows,
	collect_period_incentive_rows,
	calculation_level,
	scheme_settings,
)


def execute(filters=None):
	filters = frappe._dict(filters or {})
	period = filters.get("period") or "Monthly"
	filters.period = period
	columns = get_columns(period)
	if period in ("Monthly", "Quarterly"):
		data = collect_period_incentive_rows(filters)
	else:
		data = collect_achievement_rows(filters)
	slabs, based_on, pay_on = scheme_settings(filters)
	level = calculation_level(filters)
	data = allocate_item_incentives(data, slabs, based_on, pay_on, level)
	# Filter the completed allocation: hiding rows must never change the earned
	# slab or redistribute their incentive among the remaining visible items.
	band = filters.get("incentive_band") or "All"
	if band in ("Min", "Max"):
		data = [row for row in data if row.get("incentive_band") == band]
	elif band in ("Not Achieve", "Not Achive"):
		data = [row for row in data if not row.get("incentive_band")]
	data.sort(key=lambda row: (
		row.get("month_number") or 0, row.get("period") or "",
		row.get("sales_person") or "", row.get("territory") or "",
		row.get("item_group") or "", row.get("item_code") or "",
	))
	message = _(
		"Each row shows the item's target, sales and achievement. In Item mode, each item "
		"qualifies independently; in grouped mode, the slab is applied after combining "
		"Sales Person, Period and Customer Group. Rows at or below target receive zero. "
		"Achievement is blank when target is zero. Incentive % is the earned scheme rate used "
		"before allocation. Min/Max filter by that incentive band; Not Achieve shows rows with "
		"no allocated incentive."
	)
	if data:
		total = compute_metrics(
			sum(nflt(row.get("target_qty")) for row in data),
			sum(nflt(row.get("actual_qty")) for row in data),
			sum(nflt(row.get("target_amount")) for row in data),
			sum(nflt(row.get("actual_amount")) for row in data),
		)
		total.update({
			columns[0]["fieldname"]: _("Total"),
			"incentive_amount": sum(nflt(row.get("incentive_amount")) for row in data),
			"is_total_row": True,
		})
		data.append(total)
	# The standard footer averages percentages. Supply a total derived from
	# aggregate target/actual instead, shared by the screen and exported report.
	return columns, data, message, None, None, True


def get_columns(period="Annual"):
	qty = {"fieldtype": "Float", "precision": 0, "width": 110}
	amt = {"fieldtype": "Currency", "precision": 0, "width": 130}
	incentive_amt = {"fieldtype": "Currency", "precision": 0, "width": 150}
	pct = {"fieldtype": "Percent", "precision": 0, "width": 140}
	columns = [
		{"fieldname": "sales_person", "label": _("Sales Person"), "fieldtype": "Link", "options": "Sales Person", "width": 140},
		{"fieldname": "territory", "label": _("Territory"), "fieldtype": "Link", "options": "Territory", "width": 120},
		{"fieldname": "item_group", "label": _("Item Group"), "fieldtype": "Link", "options": "Item Group", "width": 130},
		{"fieldname": "customer_group", "label": _("Customer Group"), "fieldtype": "Link", "options": "Customer Group", "width": 140},
		{"fieldname": "item_code", "label": _("Item"), "fieldtype": "Link", "options": "Item", "width": 140},
		{"fieldname": "growth_percent", "label": _("Growth %"), **pct, "width": 110},
	]
	if period in ("Monthly", "Quarterly"):
		columns.insert(0, {"fieldname": "period", "label": _("Period"), "fieldtype": "Data", "width": 110})
	columns.extend(
		[
			{"fieldname": "target_qty", "label": _("Target Qty"), **qty},
			{"fieldname": "actual_qty", "label": _("Actual Qty"), **qty},
			{"fieldname": "variance_qty", "label": _("Qty Variance"), **qty},
			{"fieldname": "qty_achievement_percent", "label": _("Qty Achievement %"), **pct},
			{"fieldname": "target_amount", "label": _("Target Amount"), **amt},
			{"fieldname": "actual_amount", "label": _("Actual Amount"), **amt},
			{"fieldname": "variance_amount", "label": _("Amount Variance"), **amt},
			{"fieldname": "amount_achievement_percent", "label": _("Amount Achievement %"), **pct, "width": 160},
			{"fieldname": "pay_on", "label": _("Pay On"), "fieldtype": "Data", "width": 90},
			{"fieldname": "incentive_rate_percent", "label": _("Incentive %"), **pct, "width": 110},
			{"fieldname": "incentive_amount", "label": _("Allocated Incentive"), **incentive_amt},
			{"fieldname": "planning", "label": _("Planning"), "fieldtype": "Link", "options": "Sales Target Planning", "width": 140},
		]
	)
	return columns
